src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.AbstractPriorityProcessor.cs

```diff
@@ -71,31 +71,22 @@ protected async Task WaitForHigherPriorityOperationsAsync()
                     {
                         using (Logger.LogBlock(FunctionId.WorkCoordinator_WaitForHigherPriorityOperationsAsync, CancellationToken))
                         {
-                            do
+                            while (true)
                             {
-                                // Host is shutting down
-                                if (CancellationToken.IsCancellationRequested)
-                                {
-                                    return;
-                                }
+                                CancellationToken.ThrowIfCancellationRequested();
 
                                 // we wait for global operation and higher queue operation if there is anything going on
-                                if (!GlobalOperationTask.IsCompleted || !HigherQueueOperationTask.IsCompleted)
-                                {
-                                    await Task.WhenAll(GlobalOperationTask, HigherQueueOperationTask).ConfigureAwait(false);
-                                }
-
-                                // if there are no more work left for higher queue, then it is our time to go ahead
-                                if (!HigherQueueHasWorkItem)
-                                {
+                                await HigherQueueOperationTask.ConfigureAwait(false);
+
+                                // if there are no more work left for higher queue, and we didn't enter a state where we
+                                // should wait for idle again, then our time to go ahead.
+                                if (!HigherQueueHasWorkItem && !ShouldWaitForIdle())
```

Comment:
important cahnge.  while waiting for the higher pri work to be done.  we should go back to idling if other stuff happened (like us getting paused for a global operation).